File: internal/aggregate/report.go
package aggregate

import (
	"sort"

	"mcp.chic.md/internal/moysklad"
)

// This file maps raw MoySklad report rows into compact, LLM-friendly DTOs.
// Every monetary field is converted from minor to major units here. Margins are
// recomputed as profit/revenue*100 so the value is unambiguous regardless of
// how MoySklad encodes its own `margin` field.

// ---- Report envelope ------------------------------------------------------

// Report is the uniform envelope every list-style report tool returns. Totals
// are ALWAYS computed over the complete result set (every row MoySklad returned
// for the period); Rows carries only the top Returned rows, sorted by the
// report's primary money metric. Truncated tells the model that Rows is a subset
// — the totals still cover every row, so it must never re-sum Rows to answer a
// "what's the total" question. This is the fix for silent 100-row truncation:
// the model gets the correct grand total regardless of how many detail rows fit.
type Report[Row any, Totals any] struct {
	Totals    Totals `json:"totals"`
	RowCount  int    `json:"rowCount"`  // rows in the full result set
	Returned  int    `json:"returned"`  // rows included in Rows
	Truncated bool   `json:"truncated"` // RowCount > Returned
	Rows      []Row  `json:"rows"`
}

// newReport truncates rows to a display limit (<=0 means "all") and records
// whether truncation occurred. Totals are passed in already computed over the
// full slice, so truncation never affects them.
func newReport[Row any, Totals any](rows []Row, totals Totals, limit int) Report[Row, Totals] {
	full := len(rows)
	truncated := false
	if limit > 0 && limit < full {
		rows = rows[:limit]
		truncated = true
	}
	if rows == nil {
		rows = []Row{}
	}
	return Report[Row, Totals]{
		Totals:    totals,
		RowCount:  full,
		Returned:  len(rows),
		Truncated: truncated,
		Rows:      rows,
	}
}

// ---- Dashboard ------------------------------------------------------------

type DashboardSummary struct {
	Period       string  `json:"period"`
	SalesCount   int     `json:"salesCount"`
	SalesAmount  float64 `json:"salesAmount"`
	SalesDelta   float64 `json:"salesDeltaVsPrev"`
	OrdersCount  int     `json:"ordersCount"`
	OrdersAmount float64 `json:"ordersAmount"`
	MoneyIncome  float64 `json:"moneyIncome"`
	MoneyOutcome float64 `json:"moneyOutcome"`
	MoneyBalance float64 `json:"moneyBalance"`
}

// Dashboard money amounts are assumed to be in minor units (as elsewhere in
// MoySklad); verify against a real account and adjust here if a field turns out
// to be in major units.
func Dashboard(period string, d *moysklad.Dashboard) DashboardSummary {
	return DashboardSummary{
		Period:       period,
		SalesCount:   d.Sales.Count,
		SalesAmount:  MinorToMajor(d.Sales.Amount),
		SalesDelta:   MinorToMajor(d.Sales.MovementAmount),
		OrdersCount:  d.Orders.Count,
		OrdersAmount: MinorToMajor(d.Orders.Amount),
		MoneyIncome:  MinorToMajor(d.Money.Income),
		MoneyOutcome: MinorToMajor(d.Money.Outcome),
		MoneyBalance: MinorToMajor(d.Money.Balance),
	}
}

// ---- Profit ---------------------------------------------------------------

type ProfitProductLine struct {
	Name         string  `json:"name"`
	Code         string  `json:"code,omitempty"`
	SellQuantity float64 `json:"sellQuantity"`
	Revenue      float64 `json:"revenue"`
	Cost         float64 `json:"cost"`
	ReturnSum    float64 `json:"returnSum"`
	Profit       float64 `json:"profit"`
	MarginPct    float64 `json:"marginPct"`
}

func ProfitByProduct(rows []moysklad.ProfitByProductRow) []ProfitProductLine {
	out := make([]ProfitProductLine, 0, len(rows))
	for _, r := range rows {
		revenue := MinorToMajor(r.SellSum)
		profit := MinorToMajor(r.Profit)
		out = append(out, ProfitProductLine{
			Name:         r.Assortment.Name,
			Code:         r.Assortment.Code,
			SellQuantity: r.SellQuantity,
			Revenue:      revenue,
			Cost:         MinorToMajor(r.SellCostSum),
			ReturnSum:    MinorToMajor(r.ReturnSum),
			Profit:       profit,
			MarginPct:    marginPct(profit, revenue),
		})
	}
	return out
}

// ProfitTotals are the period grand totals for a profit report, over every row.
type ProfitTotals struct {
	Revenue      float64 `json:"revenue"`
	Cost         float64 `json:"cost"`
	Profit       float64 `json:"profit"`
	ReturnSum    float64 `json:"returnSum,omitempty"`
	SellQuantity float64 `json:"sellQuantity,omitempty"`
	SalesCount   int     `json:"salesCount,omitempty"`
	MarginPct    float64 `json:"marginPct"`
}

// ProfitProductReport maps rows, sorts by revenue desc, computes complete
// totals, and truncates the detail list to limit. This is what get_profit
// returns so the model never has to sum rows itself.
func ProfitProductReport(rows []moysklad.ProfitByProductRow, limit int) Report[ProfitProductLine, ProfitTotals] {
	lines := ProfitByProduct(rows)
	sort.SliceStable(lines, func(i, j int) bool { return lines[i].Revenue > lines[j].Revenue })
	var t ProfitTotals
	for _, l := range lines {
		t.Revenue += l.Revenue
		t.Cost += l.Cost
		t.Profit += l.Profit
		t.ReturnSum += l.ReturnSum
		t.SellQuantity += l.SellQuantity
	}
	t.finalize()
	return newReport(lines, t, limit)
}

// ProfitEntityReport is ProfitProductReport for the counterparty/channel/
// employee dimension.
func ProfitEntityReport(rows []moysklad.ProfitByEntityRow, limit int) Report[ProfitEntityLine, ProfitTotals] {
	lines := ProfitByEntity(rows)
	sort.SliceStable(lines, func(i, j int) bool { return lines[i].Revenue > lines[j].Revenue })
	var t ProfitTotals
	for _, l := range lines {
		t.Revenue += l.Revenue
		t.Cost += l.Cost
		t.Profit += l.Profit
		t.SalesCount += l.SalesCount
	}
	t.finalize()
	return newReport(lines, t, limit)
}

// finalize rounds accumulated sums and derives the blended margin.
func (t *ProfitTotals) finalize() {
	t.Revenue = round2(t.Revenue)
	t.Cost = round2(t.Cost)
	t.Profit = round2(t.Profit)
	t.ReturnSum = round2(t.ReturnSum)
	t.SellQuantity = round2(t.SellQuantity)
	t.MarginPct = marginPct(t.Profit, t.Revenue)
}

type ProfitEntityLine struct {
	Name       string  `json:"name"`
	Revenue    float64 `json:"revenue"`
	Cost       float64 `json:"cost"`
	Profit     float64 `json:"profit"`
	SalesCount int     `json:"salesCount"`
	AvgCheck   float64 `json:"avgCheck"`
	MarginPct  float64 `json:"marginPct"`
}

func ProfitByEntity(rows []moysklad.ProfitByEntityRow) []ProfitEntityLine {
	out := make([]ProfitEntityLine, 0, len(rows))
	for _, r := range rows {
		revenue := MinorToMajor(r.SellSum)
		profit := MinorToMajor(r.Profit)
		out = append(out, ProfitEntityLine{
			Name:       r.Name(),
			Revenue:    revenue,
			Cost:       MinorToMajor(r.SellCostSum),
			Profit:     profit,
			SalesCount: r.SalesCount,
			AvgCheck:   MinorToMajor(r.SalesAvgCheck),
			MarginPct:  marginPct(profit, revenue),
		})
	}
	return out
}

// ---- Turnover -------------------------------------------------------------

type TurnoverLine struct {
	Name       string  `json:"name"`
	StartQty   float64 `json:"startQty"`
	IncomeQty  float64 `json:"incomeQty"`
	OutcomeQty float64 `json:"outcomeQty"`
	EndQty     float64 `json:"endQty"`
	EndValue   float64 `json:"endValue"`
	// TurnoverDays is avg stock divided by average daily outbound quantity over
	// the period; 0 when there was no outbound movement (candidate dead stock).
	TurnoverDays float64 `json:"turnoverDays"`
}

// Turnover converts rows and computes turnover days using the period length in
// days (pass the number of days between momentFrom and momentTo).
func Turnover(rows []moysklad.TurnoverRow, periodDays float64) []TurnoverLine {
	out := make([]TurnoverLine, 0, len(rows))
	for _, r := range rows {
		out = append(out, TurnoverLine{
			Name:         r.Assortment.Name,
			StartQty:     r.OnPeriodStart.Quantity,
			IncomeQty:    r.Income.Quantity,
			OutcomeQty:   r.Outcome.Quantity,
			EndQty:       r.OnPeriodEnd.Quantity,
			EndValue:     MinorToMajor(r.OnPeriodEnd.Sum),
			TurnoverDays: turnoverDays(r, periodDays),
		})
	}
	return out
}

// TurnoverTotals are the period grand totals for a turnover report.
type TurnoverTotals struct {
	IncomeQty  float64 `json:"incomeQty"`
	OutcomeQty float64 `json:"outcomeQty"`
	EndValue   float64 `json:"endValue"`
}

// TurnoverReport maps rows, sorts by closing inventory value desc, totals over
// all rows, and truncates the detail list to limit.
func TurnoverReport(rows []moysklad.TurnoverRow, periodDays float64, limit int) Report[TurnoverLine, TurnoverTotals] {
	lines := Turnover(rows, periodDays)
	sort.SliceStable(lines, func(i, j int) bool { return lines[i].EndValue > lines[j].EndValue })
	var t TurnoverTotals
	for _, l := range lines {
		t.IncomeQty += l.IncomeQty
		t.OutcomeQty += l.OutcomeQty
		t.EndValue += l.EndValue
	}
	t.IncomeQty = round2(t.IncomeQty)
	t.OutcomeQty = round2(t.OutcomeQty)
	t.EndValue = round2(t.EndValue)
	return newReport(lines, t, limit)
}

func turnoverDays(r moysklad.TurnoverRow, periodDays float64) float64 {
	out := r.Outcome.Quantity
	if out <= 0 || periodDays <= 0 {
		return 0
	}
	avgStock := (r.OnPeriodStart.Quantity + r.OnPeriodEnd.Quantity) / 2
	dailyOut := out / periodDays
	if dailyOut <= 0 {
		return 0
	}
	return round2(avgStock / dailyOut)
}

// ---- Stock ----------------------------------------------------------------

type StockLine struct {
	Name       string  `json:"name"`
	Code       string  `json:"code,omitempty"`
	Article    string  `json:"article,omitempty"`
	Stock      float64 `json:"stock"`
	Reserve    float64 `json:"reserve"`
	Available  float64 `json:"available"`
	InTransit  float64 `json:"inTransit"`
	CostPrice  float64 `json:"costPrice"`
	SalePrice  float64 `json:"salePrice"`
	StockValue float64 `json:"stockValue"`
	StockDays  int     `json:"stockDays"`
}

func Stock(rows []moysklad.StockRow) []StockLine {
	out := make([]StockLine, 0, len(rows))
	for _, r := range rows {
		cost := MinorToMajor(r.Price)
		out = append(out, StockLine{
			Name:       r.Name,
			Code:       r.Code,
			Article:    r.Article,
			Stock:      r.Stock,
			Reserve:    r.Reserve,
			Available:  r.Stock - r.Reserve,
			InTransit:  r.InTransit,
			CostPrice:  cost,
			SalePrice:  MinorToMajor(r.SalePrice),
			StockValue: round2(r.Stock * cost),
			StockDays:  r.StockDays,
		})
	}
	return out
}

// StockTotals are the warehouse grand totals for a stock report — the direct
// answer to "how much money is tied up in stock".
type StockTotals struct {
	Units      float64 `json:"units"`
	Available  float64 `json:"available"`
	StockValue float64 `json:"stockValue"`
}

// StockReport maps rows, sorts by tied-up value desc, totals over all rows, and
// truncates the detail list to limit.
func StockReport(rows []moysklad.StockRow, limit int) Report[StockLine, StockTotals] {
	lines := Stock(rows)
	sort.SliceStable(lines, func(i, j int) bool { return lines[i].StockValue > lines[j].StockValue })
	var t StockTotals
	for _, l := range lines {
		t.Units += l.Stock
		t.Available += l.Available
		t.StockValue += l.StockValue
	}
	t.Units = round2(t.Units)
	t.Available = round2(t.Available)
	t.StockValue = round2(t.StockValue)
	return newReport(lines, t, limit)
}

// ---- Counterparty report --------------------------------------------------

type CounterpartyMetric struct {
	Name         string  `json:"name"`
	FirstDemand  string  `json:"firstDemand,omitempty"`
	LastDemand   string  `json:"lastDemand,omitempty"`
	DemandsCount int     `json:"demandsCount"`
	Revenue      float64 `json:"revenue"`
	AvgReceipt   float64 `json:"avgReceipt"`
	ReturnsSum   float64 `json:"returnsSum"`
	Balance      float64 `json:"balance"`
	Profit       float64 `json:"profit"`
}

func CounterpartyMetrics(rows []moysklad.CounterpartyRow) []CounterpartyMetric {
	out := make([]CounterpartyMetric, 0, len(rows))
	for _, r := range rows {
		out = append(out, CounterpartyMetric{
			Name:         r.Counterparty.Name,
			FirstDemand:  r.FirstDemandDate,
			LastDemand:   r.LastDemandDate,
			DemandsCount: r.DemandsCount,
			Revenue:      MinorToMajor(r.DemandsSum),
			AvgReceipt:   MinorToMajor(r.AverageReceipt),
			ReturnsSum:   MinorToMajor(r.ReturnsSum),
			Balance:      MinorToMajor(r.Balance),
			Profit:       MinorToMajor(r.Profit),
		})
	}
	return out
}

// CounterpartyTotals are the grand totals across all counterparties in a report.
type CounterpartyTotals struct {
	Revenue    float64 `json:"revenue"`
	Profit     float64 `json:"profit"`
	ReturnsSum float64 `json:"returnsSum,omitempty"`
	Balance    float64 `json:"balance"`
	MarginPct  float64 `json:"marginPct"`
}

// CounterpartyReport maps rows, sorts by revenue desc, totals over all rows, and
// truncates the detail list to limit.
func CounterpartyReport(rows []moysklad.CounterpartyRow, limit int) Report[CounterpartyMetric, CounterpartyTotals] {
	lines := CounterpartyMetrics(rows)
	sort.SliceStable(lines, func(i, j int) bool { return lines[i].Revenue > lines[j].Revenue })
	var t CounterpartyTotals
	for _, l := range lines {
		t.Revenue += l.Revenue
		t.Profit += l.Profit
		t.ReturnsSum += l.ReturnsSum
		t.Balance += l.Balance
	}
	t.Revenue = round2(t.Revenue)
	t.Profit = round2(t.Profit)
	t.ReturnsSum = round2(t.ReturnsSum)
	t.Balance = round2(t.Balance)
	t.MarginPct = marginPct(t.Profit, t.Revenue)
	return newReport(lines, t, limit)
}

// ---- Money ----------------------------------------------------------------

type MoneyFlow struct {
	Income  float64          `json:"income"`
	Outcome float64          `json:"outcome"`
	Net     float64          `json:"net"`
	Series  []MoneyFlowPoint `json:"series"`
}

type MoneyFlowPoint struct {
	Date    string  `json:"date"`
	Income  float64 `json:"income"`
	Outcome float64 `json:"outcome"`
	Balance float64 `json:"balance"`
}

func Money(m *moysklad.MoneySeries) MoneyFlow {
	income := MinorToMajor(m.Credit)
	outcome := MinorToMajor(m.Debit)
	pts := make([]MoneyFlowPoint, 0, len(m.Series))
	for _, p := range m.Series {
		pts = append(pts, MoneyFlowPoint{
			Date:    p.Date,
			Income:  MinorToMajor(p.Credit),
			Outcome: MinorToMajor(p.Debit),
			Balance: MinorToMajor(p.Balance),
		})
	}
	return MoneyFlow{Income: income, Outcome: outcome, Net: round2(income - outcome), Series: pts}
}

// ---- helpers --------------------------------------------------------------

func marginPct(profit, revenue float64) float64 {
	if revenue == 0 {
		return 0
	}
	return round2(profit / revenue * 100)
}

File: chic/aggregate/report.py
"""Map raw MoySklad report rows into compact DTOs (minor→major units here).

Margins are recomputed as ``profit/revenue*100`` so the value is unambiguous
regardless of how MoySklad encodes its own ``margin`` field. Every ``*_report``
returns the uniform truncating envelope: totals over the full set, rows capped.
"""

from __future__ import annotations

from chic.aggregate.envelope import truncate
from chic.aggregate.models import (
    AssortmentLine,
    AuditLine,
    CounterpartyMetric,
    CounterpartyTotals,
    DashboardSummary,
    EntityRefOut,
    MoneyFlow,
    MoneyFlowPoint,
    ProductSummary,
    ProfitEntityLine,
    ProfitProductLine,
    ProfitTotals,
    Report,
    SalesSeriesOut,
    SalesSeriesPointOut,
    StateOut,
    StockLine,
    StockTotals,
    StoreStockLine,
    StoreStockTotals,
    TurnoverLine,
    TurnoverTotals,
)
from chic.aggregate.money import dec, margin_pct, minor_to_major, money_round, round2
from chic.moysklad.models import (
    AssortmentRow,
    AuditRow,
    CounterpartyRow,
    Dashboard,
    EntityRef,
    MoneySeries,
    Product,
    ProfitByEntityRow,
    ProfitByProductRow,
    SalesSeries,
    State,
    StockByStoreRow,
    StockRow,
    TurnoverRow,
)


def product_key(meta_href: str, name: str) -> str:
    """Stable per-product join key, shared by the live reports and the snapshot
    history so their keys always match. Falls back to the name when the API omits
    the href (an empty href would otherwise collapse distinct products into one).
    """
    return meta_href or f"name:{name}"


# ---- dashboard ------------------------------------------------------------


def dashboard(period: str, d: Dashboard) -> DashboardSummary:
    return DashboardSummary(
        period=period,
        sales_count=d.sales.count,
        sales_amount=minor_to_major(d.sales.amount),
        sales_delta_vs_prev=minor_to_major(d.sales.movement_amount),
        orders_count=d.orders.count,
        orders_amount=minor_to_major(d.orders.amount),
        money_income=minor_to_major(d.money.income),
        money_outcome=minor_to_major(d.money.outcome),
        money_balance=minor_to_major(d.money.balance),
    )


# ---- product --------------------------------------------------------------


def product(p: Product) -> ProductSummary:
    return ProductSummary(
        id=p.id,
        name=p.name,
        code=p.code,
        article=p.article,
        archived=p.archived,
        sale_price=minor_to_major(p.sale_prices[0].value) if p.sale_prices else dec(0),
        buy_price=minor_to_major(p.buy_price.value) if p.buy_price is not None else dec(0),
    )


def products(ps: list[Product]) -> list[ProductSummary]:
    return [product(p) for p in ps]


# ---- profit ---------------------------------------------------------------


def profit_by_product(rows: list[ProfitByProductRow]) -> list[ProfitProductLine]:
    out: list[ProfitProductLine] = []
    for r in rows:
        revenue = minor_to_major(r.sell_sum)
        profit = minor_to_major(r.profit)
        out.append(
            ProfitProductLine(
                id=product_key(r.assortment.meta.href, r.assortment.name),
                name=r.assortment.name,
                code=r.assortment.code,
                sell_quantity=r.sell_quantity,
                revenue=revenue,
                cost=minor_to_major(r.sell_cost_sum),
                return_sum=minor_to_major(r.return_sum),
                profit=profit,
                margin_pct=margin_pct(profit, revenue),
            )
        )
    return out


def profit_by_entity(rows: list[ProfitByEntityRow]) -> list[ProfitEntityLine]:
    out: list[ProfitEntityLine] = []
    for r in rows:
        revenue = minor_to_major(r.sell_sum)
        profit = minor_to_major(r.profit)
        out.append(
            ProfitEntityLine(
                name=r.holder_name,
                revenue=revenue,
                cost=minor_to_major(r.sell_cost_sum),
                profit=profit,
                sales_count=r.sales_count,
                avg_check=minor_to_major(r.sales_avg_check),
                margin_pct=margin_pct(profit, revenue),
            )
        )
    return out


def profit_product_report(
    rows: list[ProfitByProductRow], limit: int
) -> Report[ProfitProductLine, ProfitTotals]:
    lines = sorted(profit_by_product(rows), key=lambda x: x.revenue, reverse=True)
    revenue = sum((x.revenue for x in lines), dec(0))
    profit = sum((x.profit for x in lines), dec(0))
    totals = ProfitTotals(
        revenue=money_round(revenue),
        cost=money_round(sum((x.cost for x in lines), dec(0))),
        profit=money_round(profit),
        return_sum=money_round(sum((x.return_sum for x in lines), dec(0))),
        sell_quantity=round2(sum(x.sell_quantity for x in lines)),
        margin_pct=margin_pct(profit, revenue),
    )
    shown, full, truncated = truncate(lines, limit)
    return Report[ProfitProductLine, ProfitTotals](
        totals=totals, row_count=full, returned=len(shown), truncated=truncated, rows=shown
    )


def profit_entity_report(
    rows: list[ProfitByEntityRow], limit: int
) -> Report[ProfitEntityLine, ProfitTotals]:
    lines = sorted(profit_by_entity(rows), key=lambda x: x.revenue, reverse=True)
    revenue = sum((x.revenue for x in lines), dec(0))
    profit = sum((x.profit for x in lines), dec(0))
    totals = ProfitTotals(
        revenue=money_round(revenue),
        cost=money_round(sum((x.cost for x in lines), dec(0))),
        profit=money_round(profit),
        sales_count=sum(x.sales_count for x in lines),
        margin_pct=margin_pct(profit, revenue),
    )
    shown, full, truncated = truncate(lines, limit)
    return Report[ProfitEntityLine, ProfitTotals](
        totals=totals, row_count=full, returned=len(shown), truncated=truncated, rows=shown
    )


# ---- turnover -------------------------------------------------------------


def _turnover_days(r: TurnoverRow, period_days: float) -> float:
    out = r.outcome.quantity
    if out <= 0 or period_days <= 0:
        return 0.0
    avg_stock = (r.on_period_start.quantity + r.on_period_end.quantity) / 2
    daily_out = out / period_days
    if daily_out <= 0:
        return 0.0
    return round2(avg_stock / daily_out)


def turnover(rows: list[TurnoverRow], period_days: float) -> list[TurnoverLine]:
    return [
        TurnoverLine(
            name=r.assortment.name,
            start_qty=r.on_period_start.quantity,
            income_qty=r.income.quantity,
            outcome_qty=r.outcome.quantity,
            end_qty=r.on_period_end.quantity,
            end_value=minor_to_major(r.on_period_end.sum),
            turnover_days=_turnover_days(r, period_days),
        )
        for r in rows
    ]


def turnover_report(
    rows: list[TurnoverRow], period_days: float, limit: int
) -> Report[TurnoverLine, TurnoverTotals]:
    lines = sorted(turnover(rows, period_days), key=lambda x: x.end_value, reverse=True)
    totals = TurnoverTotals(
        income_qty=round2(sum(x.income_qty for x in lines)),
        outcome_qty=round2(sum(x.outcome_qty for x in lines)),
        end_value=money_round(sum((x.end_value for x in lines), dec(0))),
    )
    shown, full, truncated = truncate(lines, limit)
    return Report[TurnoverLine, TurnoverTotals](
        totals=totals, row_count=full, returned=len(shown), truncated=truncated, rows=shown
    )


# ---- stock ----------------------------------------------------------------


def stock(rows: list[StockRow]) -> list[StockLine]:
    out: list[StockLine] = []
    for r in rows:
        cost = minor_to_major(r.price)
        out.append(
            StockLine(
                id=product_key(r.meta.href, r.name),
                name=r.name,
                code=r.code,
                article=r.article,
                stock=r.stock,
                reserve=r.reserve,
                available=r.stock - r.reserve,
                in_transit=r.in_transit,
                cost_price=cost,
                sale_price=minor_to_major(r.sale_price),
                stock_value=money_round(dec(r.stock) * cost),
                stock_days=int(r.stock_days),
            )
        )
    return out


def stock_report(rows: list[StockRow], limit: int) -> Report[StockLine, StockTotals]:
    lines = sorted(stock(rows), key=lambda x: x.stock_value, reverse=True)
    totals = StockTotals(
        units=round2(sum(x.stock for x in lines)),
        available=round2(sum(x.available for x in lines)),
        stock_value=money_round(sum((x.stock_value for x in lines), dec(0))),
    )
    shown, full, truncated = truncate(lines, limit)
    return Report[StockLine, StockTotals](
        totals=totals, row_count=full, returned=len(shown), truncated=truncated, rows=shown
    )


# ---- counterparty report --------------------------------------------------


def counterparty_metrics(rows: list[CounterpartyRow]) -> list[CounterpartyMetric]:
    return [
        CounterpartyMetric(
            name=r.counterparty.name,
            first_demand=r.first_demand_date,
            last_demand=r.last_demand_date,
            demands_count=r.demands_count,
            revenue=minor_to_major(r.demands_sum),
            avg_receipt=minor_to_major(r.average_receipt),
            returns_sum=minor_to_major(r.returns_sum),
            balance=minor_to_major(r.balance),
            profit=minor_to_major(r.profit),
        )
        for r in rows
    ]


def counterparty_report(
    rows: list[CounterpartyRow], limit: int
) -> Report[CounterpartyMetric, CounterpartyTotals]:
    lines = sorted(counterparty_metrics(rows), key=lambda x: x.revenue, reverse=True)
    revenue = sum((x.revenue for x in lines), dec(0))
    profit = sum((x.profit for x in lines), dec(0))
    totals = CounterpartyTotals(
        revenue=money_round(revenue),
        profit=money_round(profit),
        returns_sum=money_round(sum((x.returns_sum for x in lines), dec(0))),
        balance=money_round(sum((x.balance for x in lines), dec(0))),
        margin_pct=margin_pct(profit, revenue),
    )
    shown, full, truncated = truncate(lines, limit)
    return Report[CounterpartyMetric, CounterpartyTotals](
        totals=totals, row_count=full, returned=len(shown), truncated=truncated, rows=shown
    )


# ---- money flow -----------------------------------------------------------


def money(m: MoneySeries) -> MoneyFlow:
    income = minor_to_major(m.credit)
    outcome = minor_to_major(m.debit)
    return MoneyFlow(
        income=income,
        outcome=outcome,
        net=money_round(income - outcome),
        series=[
            MoneyFlowPoint(
                date=p.date,
                income=minor_to_major(p.credit),
                outcome=minor_to_major(p.debit),
                balance=minor_to_major(p.balance),
            )
            for p in m.series
        ],
    )


# ---- reference dictionaries -----------------------------------------------


def entity_refs(rows: list[EntityRef]) -> list[EntityRefOut]:
    return [
        EntityRefOut(
            id=r.id,
            name=r.name,
            code=r.code,
            archived=r.archived,
            description=r.description,
            path_name=r.path_name,
            inn=r.inn,
        )
        for r in rows
    ]


def states(rows: list[State]) -> list[StateOut]:
    return [
        StateOut(id=r.id, name=r.name, type=r.state_type, entity_type=r.entity_type) for r in rows
    ]


# ---- assortment -----------------------------------------------------------


def _assortment_kind(row: AssortmentRow) -> str:
    # meta.type is product|variant|bundle|service|consignment.
    return row.meta.type or "product"


def assortment(rows: list[AssortmentRow]) -> list[AssortmentLine]:
    return [
        AssortmentLine(
            id=r.id,
            kind=_assortment_kind(r),
            name=r.name,
            code=r.code,
            article=r.article,
            sale_price=minor_to_major(r.sale_prices[0].value) if r.sale_prices else dec(0),
            buy_price=minor_to_major(r.buy_price.value) if r.buy_price is not None else dec(0),
            stock=r.quantity,
        )
        for r in rows
    ]


# ---- stock by store -------------------------------------------------------


def stock_by_store(rows: list[StockByStoreRow]) -> Report[StoreStockLine, StoreStockTotals]:
    """Pivot per-product/per-store rows into a per-warehouse summary.

    The bystore report carries quantities but not prices, so this aggregates
    units/reserve/available per store (answering "how much sits where"), not value.
    """
    agg: dict[str, StoreStockLine] = {}
    for row in rows:
        for s in row.stock_by_store:
            line = agg.get(s.name)
            if line is None:
                line = StoreStockLine(
                    store=s.name, positions=0, units=0.0, reserve=0.0, available=0.0
                )
                agg[s.name] = line
            if s.stock != 0 or s.reserve != 0 or s.in_transit != 0:
                line.positions += 1
            line.units += s.stock
            line.reserve += s.reserve
            line.available += s.stock - s.reserve
    lines = sorted(agg.values(), key=lambda x: x.units, reverse=True)
    for line in lines:
        line.units = round2(line.units)
        line.reserve = round2(line.reserve)
        line.available = round2(line.available)
    totals = StoreStockTotals(
        stores=len(lines),
        units=round2(sum(x.units for x in lines)),
        reserve=round2(sum(x.reserve for x in lines)),
        available=round2(sum(x.available for x in lines)),
    )
    return Report[StoreStockLine, StoreStockTotals](
        totals=totals, row_count=len(lines), returned=len(lines), truncated=False, rows=lines
    )


# ---- sales / orders series ------------------------------------------------


def sales_series(kind: str, s: SalesSeries) -> SalesSeriesOut:
    points = [
        SalesSeriesPointOut(date=p.date, quantity=p.quantity, sum=minor_to_major(p.sum))
        for p in s.series
    ]
    return SalesSeriesOut(
        kind=kind,
        total_quantity=round2(sum(p.quantity for p in points)),
        total_sum=money_round(sum((p.sum for p in points), dec(0))),
        series=points,
    )


# ---- audit ----------------------------------------------------------------


def audit(rows: list[AuditRow]) -> list[AuditLine]:
    return [
        AuditLine(
            moment=r.moment,
            employee=r.uid,
            entity_type=r.entity_type,
            event_type=r.event_type,
            object_count=r.object_count,
            source=r.source,
        )
        for r in rows
    ]
